crates/red_knot/src/program/check.rs

```diff
@@ -81,212 +67,343 @@ impl Program {
+    program: &'a Program,
+
+    /// The aggregated diagnostics
+    diagnostics: Vec<String>,
 }
 
-impl<'program, 'scope_ref, 'scope> RayonCheckScheduler<'program, 'scope_ref, 'scope> {
-    pub fn new(program: &'program Program, scope: &'scope_ref rayon::Scope<'scope>) -> Self {
-        Self { program, scope }
+impl<'a> CheckContext<'a> {
+    fn new(program: &'a Program) -> Self {
+        Self {
+            scheduled_files: FxHashSet::default(),
+            program,
+            diagnostics: Vec::new(),
+        }
     }
-}
 
-impl<'program, 'scope_ref, 'scope> CheckScheduler
-    for RayonCheckScheduler<'program, 'scope_ref, 'scope>
-where
-    'program: 'scope,
-{
-    fn check_file(&self, check_file_task: CheckFileTask) {
-        let child_span =
-            tracing::trace_span!("check_file", file_id = check_file_task.file_id.as_u32());
-        let program = self.program;
+    /// Returns the tasks to check all open files in the workspace.
+    pub fn check_open_files(&mut self) -> Vec<CheckOpenFileTask> {
+        self.scheduled_files
+            .extend(self.program.workspace().open_files());
 
-        self.scope
-            .spawn(move |_| child_span.in_scope(|| check_file_task.run(program)));
+        self.program
+            .workspace()
+            .open_files()
+            .map(|file_id| CheckOpenFileTask { file_id })
+            .collect()
+    }
 
-        if current_num_threads() == 1 {
-            yield_local();
+    /// Returns the task to check a dependency.
+    pub fn check_dependency(&mut self, file_id: FileId) -> Option<CheckDependencyTask> {
+        if self.scheduled_files.insert(file_id) {
+            Some(CheckDependencyTask { file_id })
+        } else {
+            None
         }
     }
 
-    fn max_concurrency(&self) -> Option<NonZeroUsize> {
-        if current_num_threads() == 1 {
-            return None;
+    /// Pushes the result for a single file check operation
+    pub fn push_diagnostics(&mut self, diagnostics: &Diagnostics) {
+        self.diagnostics.extend_from_slice(diagnostics);
+    }
+
+    /// Returns a reference to the program that is being checked.
+    pub fn program(&self) -> &'a Program {
+        self.program
+    }
+
+    /// Creates a task context that is used to check a single file.
+    pub fn task_context<'b, S>(&self, dependency_scheduler: &'b S) -> CheckTaskContext<'a, 'b, S>
+    where
+        S: ScheduleDependency,
+    {
+        CheckTaskContext {
+            program: self.program,
+            dependency_scheduler,
         }
+    }
 
-        Some(NonZeroUsize::new(current_num_threads()).unwrap_or(NonZeroUsize::MIN))
+    fn finish(self) -> Vec<String> {
+        self.diagnostics
     }
 }
 
-/// Scheduler that runs all checks on the current thread.
-pub struct SameThreadCheckScheduler<'a> {
-    program: &'a Program,
+/// Trait that abstracts away how a dependency of a file gets scheduled for checking.
+pub trait ScheduleDependency {
```

Comment:
We could possibly make this an enum now that `program.check` only accepts an enum over two variants. However, I want to keep this a trait for now until we know how many scheduling strategies we have.